### PR TITLE
feat: add user list to bot settings

### LINKED ISSUE #76: Add a user list in bot settings

### PR DESCRIPTION
**Pull request by spaenleh**
This PR adds the ability to create a list of users that will see/not see a bot's comments.
The list can be set to "see by default" or "hide by default" so that new users will get the desired visibility without needing to update the list.

Screenshot 2022-03-14 at 17 40 56

closes #76

### CODE REVIEW

File: src/components/modes/teacher/AvatarDialog.js [hunk 1]
```diff
@@ -241,6 +300,13 @@ class AvatarDialog extends Component {
     this.handlePersonalityVerification();
   };
 
+  handleChangeBotVisibility = (value) => {
+    console.log(value);
```
Comment (juancarlosfarah) [inline src/components/modes/teacher/AvatarDialog.js:304]: Remove log.

File: src/components/modes/teacher/AvatarDialog.js [hunk 2]
```diff
@@ -307,9 +373,44 @@ class AvatarDialog extends Component {
...
+    const currentIndex = userList.indexOf(value);
+    // make a copy of the array
+    const newUserList = [...userList];
+
+    // not found
+    if (currentIndex === -1) {
+      newUserList.push(value);
+    } else {
+      newUserList.splice(currentIndex, 1);
+    }
```
Comment (juancarlosfarah) [inline src/components/modes/teacher/AvatarDialog.js:388]: Maybe explicitly add a comment when the user is removed and when the user is added, for clarity.

**Review by juancarlosfarah — CHANGES_REQUESTED** [review]
Looking good! Just some minor changes. Also, I wanted to ask if you needed to add anything in the settings reducers. I noticed that because of the conditional in line 606 of CodeReview.js, it should not "hide" if the setting is null, but it might be nice to have the default settings well defined in the reducer.

File: src/components/modes/teacher/AvatarDialog.js [hunk 3]
```diff
@@ -416,80 +547,176 @@ class AvatarDialog extends Component {
...
                 >
-                  {t(validator.message)}
-                </Alert>
-              </Grid>
-            </>
-          ) : null}
-        </Grid>
+                  <Grid item>{checkAllButton}</Grid>
+                  <Grid item>{unCheckAllButton}</Grid>
+                  {/* <Grid item>{fileUploadControl}</Grid> */}
```
[the remark this answers is not in the document]
Comment (spaenleh) [reply src/components/modes/teacher/AvatarDialog.js:660]: Copy-Pasting and forgot to remove

**Comment by spaenleh** [comment]
@juancarlosfarah I added a merging step before saving the bot settings that will ensure that untouched properties (which were given a default value but were never actually saved in the state get added with their default value before saving the resource).

**Review by juancarlosfarah — APPROVED** [review]
🚀

**Pull request merged**